authentication-scheme.mediawiki

```diff
@@ -0,0 +1,221 @@
+In this case, the client should first synchronize the time or prompt the user.
+
+=====callback=====
+
+''Optional''. An url that used to send the response payload.
+
+When a <code>callback</code> is provided, the response payload is sent to the <code>callback</code> url in HTTP POST method.
+
+====Response payload====
+
+<pre>
+{
+  "algorithm": "ECDSA-P256",
+  "pubkey": "0355912bc4e61c9715c5912397ea53a5ac6c103c4893fbd9c2a9f3be13b7a3e29d",
+  "address": "NfMFWYxaUUQy9SYo6AhRiGmRxfPxe9Edj7",
+  "nonce": "13458238842203010919",
+  "timestamp": 1616131369,
+  "signature": "BAS7Ljufj3vrhOrTAi21D/5Cf62n4r64Suf/do8dq/OCMHiLJl+hLJeMFZwTajVjhcpFLz6FuSEp13vvEqWf1w=="
+}
+</pre>
+
+=====algorithm=====
+
+The algorithm used by the client to sign the response data.
+It should be one of the values in the <code>allowed_algorithms</code> sent by the server.
+
+If the algorithm used by the client is not supported by the server, the authentication fails.
+
+=====pubkey=====
+
+The public key used to verify the signature.
+It must correspond to the <code>address</code>, otherwise the authentication fails.
+
+=====address=====
+
+The address of the user to be authenticated.
+It must correspond to the <code>pubkey</code>, otherwise the authentication fails.
+
+=====nonce=====
+
+A 64-bit unsigned integer used to identify this authentication.
+This value must be exactly the same as the <code>nonce</code> sent by the server.
+
+When the server receives the response, it should check whether the nonce has expired.
+If the nonce has expired or not exists, the authentication fails.
+
+=====timestamp=====
+
+The timestamp of the client.
+This timestamp is used by the server to determine the time of the signature to prevent replay attacks.
+
+=====signature=====
+
+The signature of the response data.
+It should be encoded in base64.
+
+====Response data to sign====
+
+{| class="wikitable sortable" style="width: auto; text-align: center; font-size: smaller; table-layout: fixed;"
+!Name
+!Type
+!Offset
+!Size
+!Description
+|-
+| network
+| uint32
+| 0
+| 4
+| The magic number of the network
+|-
+| nonce
+| uint64
+| 4
+| 8
+| The nonce sent by the server
+|-
+| timestamp
+| uint32
+| 12
+| 4
+| The timestamp of the client
+|-
+| hash
+| uint160
+| 16
+| 20
+| The script hash of the user's address
+|}
+
+===Scene modes===
+
+====QR code mode====
+
+A website can allow NEO users to log in directly.
+
+The login process is:
+
+#The user clicks the login button with the NEO icon.
+
+#The QR code for login is displayed on the webpage. The challenge payload is encoded in the QR code. In this case, the <code>callback</code> must be included.
+
+#The user opens the mobile wallet and scans the QR code.
+
+#After the wallet recognizes the QR code, a pop-up window asks the user to confirm login.
+
+#If the user cancels the login, the authentication process ends. Otherwise, go to the next step.
```

Comment:
How the server gets the authentication refused message? Add `State` in Response data, include scaned, confirmed, refused, to optimize user experience?

Reply:
If the authentication is cancelled, the server won't receive the response payload.